packages/client/src/perma-cache.js

```diff
@@ -0,0 +1,396 @@
+ * TS for out-of-the box type inference and better IntelliSense.
+ *
+ * @example
+ * ```js
+ * import { PermaCache } from 'nftstorage.link'
+ *
+ * const urls = [
+ *  'https://bafkreidyeivj7adnnac6ljvzj2e3rd5xdw3revw4da7mx2ckrstapoupoq.ipfs.nftstorage.link'
+ * ]
+ *
+ * const client = new PermaCache({ token: 'YOUR_NFT_STORAGE_TOKEN' })
+ * const permaCacheEntries = await client.put(urls)
+ * await client.delete(urls)
+ * ```
+ * @module
+ */
+
+import throttledQueue from 'throttled-queue'
+import pSettle from 'p-settle'
+import pRetry, { AbortError } from 'p-retry'
+
+import { fetch } from './platform.js'
+
+const MAX_RETRIES = 5
+// These match what is enforced server-side
+const RATE_LIMIT_REQUESTS = 100
+const RATE_LIMIT_PERIOD = 60 * 1000
+const ABORTABLE_ERRORS_STATUS = [
+  400, // Invalid URL
+  401, // Not authenticated
+  403, // Not authorized
+  417, // Expectation Failed Error
+]
+
+/**
+ * @typedef { import('./lib/interface.js').RateLimiter } RateLimiter
+ * @typedef { import('./lib/interface.js').Service } Service
+ * @typedef { import('./lib/interface.js').PutOptions} PutOptions
+ * @typedef { import('./lib/interface.js').DeleteOptions} DeleteOptions
+ * @typedef { import('./lib/interface.js').ListOptions} ListOptions
+ * @typedef { import('./lib/interface.js').CacheResult} CacheResult
+ * @typedef { import('./lib/interface.js').CacheDeleteResult} CacheDeleteResult
+ * @typedef { import('./lib/interface.js').CacheEntry} CacheEntry
+ * @typedef { import('./lib/interface.js').AccountInfo} AccountInfo
+ */
+
+/**
+ * @implements Service
+ */
+export class PermaCache {
+  /**
+   * Constructs a client bound to the given `options.token` and
+   * `options.endpoint`.
+   *
+   * @example
+   * ```js
+   * import { PermaCache } from 'nftstorage.link'
+   * const client = new PermaCache({ token: API_TOKEN })
+   * ```
+   *
+   * @param {{token: string, endpoint?:URL, rateLimiter?: RateLimiter}} options
+   */
+  constructor({
+    token,
+    endpoint = new URL('https://api.nftstorage.link'),
+    rateLimiter,
+  }) {
+    /**
+     * Authorization token.
+     *
+     * @readonly
+     */
+    this.token = token
+    /**
+     * Service API endpoint `URL`.
+     * @readonly
+     */
+    this.endpoint = endpoint
+    /**
+     * @readonly
+     */
+    this.rateLimiter = rateLimiter || createRateLimiter()
+  }
+
+  /**
+   * @hidden
+   * @param {string} token
+   * @returns {Record<string, string>}
+   */
+  static headers(token) {
+    /* c8 ignore next 1 */
+    if (!token) throw new Error('missing token')
+    return {
+      Authorization: `Bearer ${token}`,
+      'X-Client': 'nftstorage.link/js',
+    }
+  }
+
+  /**
+   * @param {Service} service
+   * @param {string[]} urls
+   * @param {PutOptions} [options]
+   * @returns {Promise<CacheResult[]>}
+   */
+  static async put(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    urls,
+    { onPut, maxRetries = MAX_RETRIES } = {}
```

Comment:
There's no point adding a default value to the destructuring since you already have to check for a falsey value later when used in the case that no options object is passed.

```suggestion
    { onPut, maxRetries } = {}
```